omod/src/main/java/org/openmrs/module/spa/servlet/SpaResourcesServlet.java

```diff
@@ -0,0 +1,118 @@
+/**
+ * This Source Code Form is subject to the terms of the Mozilla Public License,
+ * v. 2.0. If a copy of the MPL was not distributed with this file, You can
+ * obtain one at http://mozilla.org/MPL/2.0/. OpenMRS is also distributed under
+ * the terms of the Healthcare Disclaimer located at http://openmrs.org/license.
+ *
+ * Copyright (C) OpenMRS Inc. OpenMRS is a registered trademark and the OpenMRS
+ * graphic logo is a trademark of OpenMRS Inc.
+ */
+package org.openmrs.module.spa.servlet;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.openmrs.api.AdministrationService;
+import org.openmrs.api.context.Context;
+import org.openmrs.module.spa.utils.SpaModuleUtils;
+import org.openmrs.util.OpenmrsUtil;
+
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+
+/**
+ * This is a servlet class that handles all requests for the frontend static resources
+ * All requests with /frontend are handled by the class
+ */
+public class SpaResourcesServlet extends HttpServlet {
+
+	private static final long serialVersionUID = 1239820102030344L;
+
+	private Log log = LogFactory.getLog(this.getClass());
+
+	/**
+	 * Used for caching purposes
+	 *
+	 * @see HttpServlet#getLastModified(HttpServletRequest)
+	 */
+	@Override
+	protected long getLastModified(HttpServletRequest req) {
+		File f = getFile(req);
+		
+		if (f == null) {
+			return super.getLastModified(req);
+		}
+		
+		return f.lastModified();
+	}
+	
+	@Override
+	protected void doGet(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
+		log.debug("In service method for spa module frontend resource servlet: " + request.getPathInfo());
+		
+		File f = getFile(request);
+		if (f == null) {
+			response.setStatus(HttpServletResponse.SC_NOT_FOUND);
+			return;
+		}
+		
+		response.setDateHeader("Last-Modified", f.lastModified());
+		response.setContentLength(Long.valueOf(f.length()).intValue());
+		String mimeType = getServletContext().getMimeType(f.getName());
+		response.setContentType(mimeType);
+		
+		FileInputStream is = new FileInputStream(f);
+		try {
+			OpenmrsUtil.copyFile(is, response.getOutputStream());
+		}
+		finally {
+			OpenmrsUtil.closeStream(is);
+		}
+	}
+	
+	/**
+	 * Turns the given request/path into a File object
+	 *
+	 * @param request the current http request
+	 * @return the file being requested or null if not found
+	 */
+	protected File getFile(HttpServletRequest request) {
+		
+		String path = request.getPathInfo(); // all url will have a base of /spa/spaModuleResources/
+
+		// we want to extract everything after /spa/spaModuleResources/ from the path info. This should cater for sub-directories
```

Comment:
I thought we were serving things at `/openmrs/spa/frontend`?

Reply:
The URL is modified by the filter. For some reasons, it gets redirected to the module servlet, the /ms/. Honestly, I just borrowed from existing code. @dkayiwa could you please provide some insight.

Reply:
@ojwanganto how do you intend to access these resources? Can you give an example url?

Reply:
@dkayiwa an example would be localhost:8080/openmrs/frontend/importmap.json

Reply:
Checking if it is really required

Reply:
I thought we were using localhost:8080/openmrs/spa/...